REGRESSION(269745@main): Inserting a PathSeg with NaN coordinates into a <path> element can result in NaN StrokeBoundingBox

shallawa opened the pull request:
#### e3aa940b07c165276a0e1f416ff92c45387da303
<pre>
REGRESSION(269745@main): Inserting a PathSeg with NaN coordinates into a &lt;path&gt; element can result in NaN StrokeBoundingBox
<a href="https://bugs.webkit.org/show_bug.cgi?id=297730">https://bugs.webkit.org/show_bug.cgi?id=297730</a>
<a href="https://rdar.apple.com/157533330">rdar://157533330</a>

Reviewed by Simon Fraser.

A pathSeg with NaN cooridatnes and Linecap should not make the StrokeBoundingBox
of an SVGPathElement a NaN FloatRect.

* LayoutTests/svg/custom/path-seg-nan-coordinates-linecap-crash-expected.txt: Added.
* LayoutTests/svg/custom/path-seg-nan-coordinates-linecap-crash.html: Added.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):

Canonical link: <a href="https://commits.webkit.org/299291@main">https://commits.webkit.org/299291@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04ff20dc085fe80c3304ed287a1426bce508c0bc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118301 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37981 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28625 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124464 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70352 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/7933075e-8590-4d2a-9770-f45cfb307e48) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120179 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38676 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46563 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89780 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59405 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4814b415-333e-4f8e-85a9-a9704ddb6b8e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121254 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30812 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106066 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70269 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ef8e535f-c3d3-496c-a1f7-8292f9d70cc3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29876 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24183 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68127 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100239 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24363 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127536 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45207 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34084 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98456 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45570 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102285 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98242 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25010 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43641 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21628 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41683 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45077 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50753 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44540 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47884 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46227 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
```diff
@@ -87,15 +87,16 @@ FloatRect LegacyRenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLin
-        auto markerRect = this->markerRect(repaintRectCalculation, strokeWidth);
-        if (!markerRect.isNaN())
-            strokeBoundingBox.unite(markerRect);
+        if (Markable<FloatRect> markerRect = this->markerRect(repaintRectCalculation, strokeWidth))
```
I think this is abuse of `Markable<FloatRect>`. You're using the fact that `Markable<FloatRect>` happens to use NaN for the invalid value, but it's the right fix to have `markerRect` return a `std::optional<FloatRect>`?

shallawa replied:
I think the check `markerRect.isNaN()` was added mainly because this traits
```
template<>
struct MarkableTraits<WebCore::FloatRect> {
    constexpr static bool isEmptyValue(const WebCore::FloatRect& rect)
    {
        return rect.isNaN();
    }
};
```
have the same check to know whether  `Markable<FloatRect>`  has a value or not. If we decide to use a different check in the future like this one for example:
```
template<>
struct MarkableTraits<WebCore::FloatRect> {
    constexpr static bool isEmptyValue(const WebCore::FloatRect& rect)
    {
        return rect.isInfinite();
    }
};
```
then we have to come here and make this change:
```
        if (!markerRect.isInfinite())
            strokeBoundingBox.unite(markerRect);
```
So putting `markerRect` in `Markable<FloatRect>` will guarantee the same check will be used here and in this function:
```
FloatRect LegacyRenderSVGShape::strokeBoundingBox() const
{
    if (!m_strokeBoundingBox)
        m_strokeBoundingBox = calculateStrokeBoundingBox();
    return *m_strokeBoundingBox;
}
```

shallawa replied:
> but it's the right fix to have `markerRect` return a `std::optional<FloatRect>`?

Using `Markable<FloatRect>` and `std::optional<FloatRect>` in this context will make things more complicated. We tend to keep all the numbers as they are especially when they are set the web developer like this test case:
```
        let pathSeg = path.createSVGPathSegMovetoAbs(NaN, NaN);
        pathSegList.replaceItem(pathSeg, 1);
```
The code was working without any check before adding `Markable<FloatRect>`. After adding it we now have to check `isNaN()` otherwise the caching state of the `m_strokeBoundingBox` will be confused.

Also what condition will make us return `std::nullopt` from `markerRect()`? Should we use `isNaN()`? How about `isInfinite()`? And if it is `isNaN()` why should not `markerRect()` return `Markable<FloatRect>` then?

shallawa committed:
REGRESSION(269745@main): Inserting a PathSeg with NaN coordinates into a <path> element can result in NaN StrokeBoundingBox
https://bugs.webkit.org/show_bug.cgi?id=297730
rdar://157533330

Reviewed by Simon Fraser.

A pathSeg with NaN cooridatnes and Linecap should not make the StrokeBoundingBox
of an SVGPathElement a NaN FloatRect.

* LayoutTests/svg/custom/path-seg-nan-coordinates-linecap-crash-expected.txt: Added.
* LayoutTests/svg/custom/path-seg-nan-coordinates-linecap-crash.html: Added.
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGPath.cpp:
(WebCore::LegacyRenderSVGPath::adjustStrokeBoundingBoxForMarkersAndZeroLengthLinecaps const):

Canonical link: https://commits.webkit.org/299291@main
